src/main/scala/org/broadinstitute/hail/driver/AddKeyTable.scala

```diff
@@ -0,0 +1,103 @@
+package org.broadinstitute.hail.driver
+
+import org.broadinstitute.hail.expr._
+import org.broadinstitute.hail.methods.Aggregators
+import org.broadinstitute.hail.utils._
+import org.kohsuke.args4j.{Option => Args4jOption}
+
+object AddKeyTable extends Command {
+
+  class Options extends BaseOptions with TextTableOptions {
+    @Args4jOption(required = true, name = "-k", aliases = Array("--key-cond"),
+      usage = "Named key condition", metaVar = "EXPR")
+    var keyCond: String = _
+
+    @Args4jOption(required = true, name = "-a", aliases = Array("--agg-cond"),
+      usage = "Named aggregation condition", metaVar = "EXPR")
+    var aggCond: String = _
+
+    @Args4jOption(required = true, name = "-o", aliases = Array("--output"),
+      usage = "output file")
+    var output: String = _
+  }
+
+  def newOptions = new Options
+
+  def name = "addkeytable"
+
+  def description = "Creates a new key table with key(s) determined by named expressions and additional columns determined by named aggregator expressions"
+
+  def supportsMultiallelic = true
+
+  def requiresVDS = true
+
+  override def hidden = true
+
+  def run(state: State, options: Options): State = {
+
+    val vds = state.vds
+    val sc = state.sc
+
+    val aggCond = options.aggCond
+    val keyCond = options.keyCond
+
+    val aggregationEC = EvalContext(Map(
+      "v" -> (0, TVariant),
+      "va" -> (1, vds.vaSignature),
+      "s" -> (2, TSample),
+      "sa" -> (3, vds.saSignature),
+      "global" -> (4, vds.globalSignature)))
+
+    val symTab = Map(
+      "v" -> (0, TVariant),
+      "va" -> (1, vds.vaSignature),
+      "s" -> (2, TSample),
+      "sa" -> (3, vds.saSignature),
+      "global" -> (4, vds.globalSignature),
+      "gs" -> (-1, BaseAggregable(aggregationEC, TGenotype)))
+
+    val ec = EvalContext(symTab)
+    val a = ec.a
+
+    ec.set(4, vds.globalAnnotation)
+    aggregationEC.set(4, vds.globalAnnotation)
+
+    val (keyNames, keyParseTypes, keyF) = Parser.parseNamedArgs(keyCond, ec)
+    val (aggNames, aggParseTypes, aggF) = Parser.parseNamedArgs(aggCond, ec)
+
+    if (keyNames.isEmpty)
+      fatal("this module requires one or more named expr arguments as keys")
+    if (aggNames.isEmpty)
+      fatal("this module requires one or more named expr arguments to aggregate by key")
```

Comment:
Are these strictly necessary?  Empty keys means the key is an empty struct and that allows you to do global aggregation.  (A scalar is the same as a single-row table.)  Empty aggregator is less obviously useful, but maybe you are just interested in the set of aggregation keys that appear...